[Experimental][ONNX] Export with symbolic shapes in proto

BowenBao opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #114767
* __->__ #112179

Experimental feature to store symbolic shapes produced by torch dynamo inside the exported onnx model.
There is no official ONNX spec to support nodes within FunctionProto to have value info, https://github.com/onnx/onnx/issues/5487. The names for value info are generated uniquely to be retrievable based on the call site and call stack.
This requires onnxscript with https://github.com/microsoft/onnxscript/tree/bowbao/export_symbolic_shapes

BowenBao committed:
[Experimental][ONNX] Export with symbolic shapes in proto

[ghstack-poisoned]

BowenBao committed:
Update on "[Experimental][ONNX] Export with symbolic shapes in proto"




[ghstack-poisoned]

BowenBao committed:
Update on "[Experimental][ONNX] Export with symbolic shapes in proto"




[ghstack-poisoned]

BowenBao commented on torch/onnx/_internal/fx/fx_onnx_interpreter.py:
```diff
@@ -243,20 +243,15 @@ def _fill_tensor_shape_type(
             continue
         elif fx_type_utils.is_torch_symbolic_type(expected_value):
             # aten::sym_size output is a int, not a tensor, which stands
-            # for the size of one dim. We treat it as 0-D tensor.
```
Previous comment was contradicting the torchlib implementation.

BowenBao committed:
Update onnxscript pin: fix beartype on "[Experimental][ONNX] Export with symbolic shapes in proto"


Experimental feature to store symbolic shapes produced by torch dynamo inside the exported onnx model.
There is no official ONNX spec to support nodes within FunctionProto to have value info, https://github.com/onnx/onnx/issues/5487. The names for value info are generated uniquely to be retrievable based on the call site and call stack.
This requires onnxscript with https://github.com/microsoft/onnxscript/tree/bowbao/export_symbolic_shapes



[ghstack-poisoned]

BowenBao committed:
Update on "[Experimental][ONNX] Export with symbolic shapes in proto"


Experimental feature to store symbolic shapes produced by torch dynamo inside the exported onnx model.
There is no official ONNX spec to support nodes within FunctionProto to have value info, https://github.com/onnx/onnx/issues/5487. The names for value info are generated uniquely to be retrievable based on the call site and call stack.
This requires onnxscript with https://github.com/microsoft/onnxscript/tree/bowbao/export_symbolic_shapes



[ghstack-poisoned]

BowenBao committed:
rebase and switch to onnxscript main branch on "[Experimental][ONNX] Export with symbolic shapes in proto"


Experimental feature to store symbolic shapes produced by torch dynamo inside the exported onnx model.
There is no official ONNX spec to support nodes within FunctionProto to have value info, https://github.com/onnx/onnx/issues/5487. The names for value info are generated uniquely to be retrievable based on the call site and call stack.
This requires onnxscript with https://github.com/microsoft/onnxscript/tree/bowbao/export_symbolic_shapes



[ghstack-poisoned]

titaiwangms commented:
> This requires onnxscript with https://github.com/microsoft/onnxscript/tree/bowbao/export_symbolic_shapes

The link doesn't work on my side. Is it a PR in onnxscript?

titaiwangms commented:
> > This requires onnxscript with https://github.com/microsoft/onnxscript/tree/bowbao/export_symbolic_shapes
> 
> The link doesn't work on my side. Is it a PR in onnxscript?

Found it. It's merged.

titaiwangms commented on torch/onnx/_internal/fx/fx_onnx_interpreter.py:
```diff
@@ -266,10 +261,9 @@ def _fill_tensor_shape_type(
-            onnxscript_value.shape = tuple(
-                [dim if isinstance(dim, int) else None for dim in expected_value.size()]
-            )
+            onnxscript_value.shape = expected_value.size()
```
I recall we avoid introducing FakeTensor concept to onnxscript, but here, we are shiping torchSym family concept to onnxscript. Is it because torch.Sym is public, while fake tensor are not?

BowenBao replied:
I wrote it like this to keep the sym handling logic in one place.

BowenBao committed:
update fix in onnxscript on "[Experimental][ONNX] Export with symbolic shapes in proto"


Experimental feature to store symbolic shapes produced by torch dynamo inside the exported onnx model.
There is no official ONNX spec to support nodes within FunctionProto to have value info, https://github.com/onnx/onnx/issues/5487. The names for value info are generated uniquely to be retrievable based on the call site and call stack.
This requires onnxscript with https://github.com/microsoft/onnxscript/tree/bowbao/export_symbolic_shapes



[ghstack-poisoned]

BowenBao committed:
rebase on "[Experimental][ONNX] Export with symbolic shapes in proto"


Experimental feature to store symbolic shapes produced by torch dynamo inside the exported onnx model.
There is no official ONNX spec to support nodes within FunctionProto to have value info, https://github.com/onnx/onnx/issues/5487. The names for value info are generated uniquely to be retrievable based on the call site and call stack.
This requires onnxscript with https://github.com/microsoft/onnxscript/tree/bowbao/export_symbolic_shapes



[ghstack-poisoned]

titaiwangms commented:
It looks like https://github.com/microsoft/onnxscript/pull/1172 breaks `test_gpt2_tiny_from_config` for some reasons. I got ShapeTypeInference Error on Concat node in ORT if I update the onnxscript version.

BowenBao committed:
rebase + switch back to onnxscript package pin on "[Experimental][ONNX] Export with symbolic shapes in proto"


Experimental feature to store symbolic shapes produced by torch dynamo inside the exported onnx model.
There is no official ONNX spec to support nodes within FunctionProto to have value info, https://github.com/onnx/onnx/issues/5487. The names for value info are generated uniquely to be retrievable based on the call site and call stack.
This requires onnxscript with https://github.com/microsoft/onnxscript/tree/bowbao/export_symbolic_shapes



[ghstack-poisoned]

BowenBao committed:
Remove xfail for passed tests on "[Experimental][ONNX] Export with symbolic shapes in proto"


Experimental feature to store symbolic shapes produced by torch dynamo inside the exported onnx model.
There is no official ONNX spec to support nodes within FunctionProto to have value info, https://github.com/onnx/onnx/issues/5487. The names for value info are generated uniquely to be retrievable based on the call site and call stack.
This requires onnxscript with https://github.com/microsoft/onnxscript/tree/bowbao/export_symbolic_shapes



[ghstack-poisoned]

BowenBao committed:
Update xfail message for input mutation failures on "[Experimental][ONNX] Export with symbolic shapes in proto"


Experimental feature to store symbolic shapes produced by torch dynamo inside the exported onnx model.
There is no official ONNX spec to support nodes within FunctionProto to have value info, https://github.com/onnx/onnx/issues/5487. The names for value info are generated uniquely to be retrievable based on the call site and call stack.
This requires onnxscript with https://github.com/microsoft/onnxscript/tree/bowbao/export_symbolic_shapes



[ghstack-poisoned]

BowenBao commented on test/onnx/test_fx_to_onnx_with_onnxruntime.py:
```diff
@@ -515,9 +515,15 @@ def forward(self, x):
-        "of node (n0__4) does not match expected type (tensor(int64))"
+    @pytorch_test_common.xfail_if_model_type_is_exportedprogram(
+        "RuntimeError:"
+        "Found following user inputs located at [0] are mutated. This is currently banned in the aot_export workflow."
```
thx for heads up. Still failing, updated error message in xfail reason.

BowenBao commented:
@pytorchbot merge

BowenBao commented:
@pytorchbot merge

titaiwangms commented:
@BowenBao You have to rebase first. I think my last PR also modifies the error message.

BowenBao committed:
rebase on "[Experimental][ONNX] Export with symbolic shapes in proto"


Experimental feature to store symbolic shapes produced by torch dynamo inside the exported onnx model.
There is no official ONNX spec to support nodes within FunctionProto to have value info, https://github.com/onnx/onnx/issues/5487. The names for value info are generated uniquely to be retrievable based on the call site and call stack.
This requires onnxscript with https://github.com/microsoft/onnxscript/tree/bowbao/export_symbolic_shapes



[ghstack-poisoned]

BowenBao commented:
@pytorchbot merge

BowenBao commented:
@pytorchbot merge

BowenBao commented:
@pytorchbot merge

thiagocrepaldi commented:
Was https://github.com/microsoft/onnxscript/tree/bowbao/export_symbolic_shapes ever merged into main?
